W-14392365 Files for editing (#245)
mulesoft/docs-code-builder: modules/ROOT/pages/_editing/int-ai-gen-flow-examples.adoc

> = Generative Flow Examples

Reviewer: Gerund? Using Generative Flow Examples? (Because the H2s are imperative.)

Feel free to ignore 🤣

Use these example prompts to help you get started with the generative flow feature and understand how to create prompts for your own use cases.  

> == New Case Notification

Reviewer: Should this be an imperative? Implement New Case Notification?

> Say your company receives numerous customer inquiries and support requests on a daily basis. You want to implement email notifications whenever a new case is created in Salesforce so you can respond quickly to customers and ensure timely case resolution to increase customer satisfaction.

Reviewer: I sort of hate this construction "Say yadda yadda", but I don't know why. I wonder how it translates.

> Say your company receives numerous customer inquiries and support requests on a daily basis. You want to implement email notifications whenever a new case is created in Salesforce so you can respond quickly to customers and ensure timely case resolution to increase customer satisfaction.

Reviewer: ```suggestion
Say your company receives numerous customer inquiries and support requests daily. You want to implement email notifications whenever a new case is created in Salesforce so you can respond quickly to customers and ensure timely case resolution to increase customer satisfaction.
```

To generate a flow for this task, enter a prompt like: `Generate a flow that sends an email when a case is created in Salesforce`.

> This generates XML code for the new flow, similar to this:  

Reviewer: ```suggestion
This prompt generates XML code for the new flow, similar to this:  
```

    </salesforce:new-object-listener>
    <salesforce:query-result-list doc:name="Query result list" doc:id="79da7485-339c-4ab6-956a-0684210b2ded" config-ref="Salesforce_Sfdc_config"/>
    <salesforce:query-result-stream-by-id doc:name="Query result stream by id" doc:id="bdf16fc8-017f-439b-b1fc-3bff23d881d8" config-ref="Salesforce_Sfdc_config"/>
    <ee:transform doc:name="Transform Message"  >
        <ee:message > <2>
           <ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
payload map {
    CaseNumber: $.CaseNumber,
    Subject: $.Subject,
    Description: $.Description,
    Status: $.Status,
    Priority: $.Priority
}]]></ee:set-payload> 
        </ee:message>
    </ee:transform>
    <email:send config-ref="Email_SMTP" fromAddress="user1@domain.com" subject="New Case Created"> <3>
        <email:to-addresses >
            <email:to-address value="user3@domain.com" />
            <email:to-address value="user2@domain.com" />
        </email:to-addresses>
        <email:body contentType="text/html" > <4>
            <email:content ><![CDATA["<h1>New Case Created:</h1>"]]></email:content>
        </email:body>
    </email:send>
</flow>
----

> <1> This flow starts with a Salesforce Connector `on-new-object-listener` so that when a new case is created, the flow is triggered. 

Reviewer: `[calloutlist]` ?

> <3> The payload for the new case is set in variables using a DataWeave script within the `transform` component. 

Reviewer: ```suggestion
<3> The payload for the new case is set in variables using a DataWeave script within the Transform component. 
```

> <4> The email is sent using the Email Connector Send operation with the configured SMTP settings.

Reviewer: ```suggestion
<4> The email is sent using the Email Connector *Send* operation with the configured SMTP settings.
```

Reviewer: https://confluence.internal.salesforce.com/display/MTDT/MuleSoft+CX+Writing+Style+Reference#MuleSoftCXWritingStyleReference-fontchoices

Reviewer: I think we need a cheatsheet for these things. I started one with Duke, but didn't finish...

+

> The subject of the email is set to `"New Case Created: "` followed by the case subject.

Reviewer: remove the " " (unless they're part of the subject)?

<5> The body of the email is set to HTML format and includes the case ID, subject, and description.

=== Complete the Flow

After reviewing and inserting the generated flow into your project:

. Type the target email address to use in your use case, for example:
+ 
`<email:to-address value="\demo@gmail.com" />`

> . Add the target email, `${email.username}` to your xref:int-create-secure-configs#properties.adoc[configuration properties file]. 

Reviewer: ```suggestion
. Add the target email, `${email.username}` to your xref:int-create-secure-configs.adoc#properties[configuration properties file]. 
```

            </scheduling-strategy>
        </salesforce:new-object-listener>
        <set-variable variableName="caseNumber" value="#[payload.CaseNumber]" doc:name="caseNumberVar" doc:id="vysmen" />
        <set-variable variableName="status" value="#[payload.Status]" doc:name="statusVar" doc:id="yeakie" />
        <email:send config-ref="Email_SMTP" doc:name="Send Email" subject="#['Case ' ++ vars.caseNumber ++ ' was created']">
            <email:to-addresses>
                <email:to-address value="demo@gmail.com" />
            </email:to-addresses>
            <email:body contentType="text/html">
                <email:content>
                    <![CDATA[#['New case ' ++ vars.caseNumber ++ ' was created with status ' ++ vars.status]]]>
                </email:content>
            </email:body>
        </email:send>
    </flow>
</mule>
----

== Update a Phone Number

Provide a streamlined process for Salesforce System Administrators by using an API call to efficiently and securely update a phone number.

To generate a flow for this task, enter a prompt like: `Create a flow that listens for an API call and updates the phone number of a Salesforce Account object`.

> This generates XML code for the new flow, similar to this:  

Reviewer: ```suggestion
This prompt generates XML code for the new flow, similar to this:  
```

[source, xml]
----
<flow name="updatePhoneNumber" doc:id="1">
    <http:listener doc:name="Listener" doc:id="2" config-ref="HTTP_Listener_config" path="/updatePhoneNumber"/> <1>
    <ee:transform doc:name="Set Payload" doc:id="3"> <2>
        <ee:message>
            <ee:set-payload><![CDATA[%dw 2.0
output application/java
---
{
    "accountId": payload.accountId,
    "phoneNumber": payload.phoneNumber
}]]></ee:set-payload> 
        </ee:message>
    </ee:transform>
    <salesforce:update doc:name="Update Account" doc:id="4" config-ref="Salesforce_Config" type="Account"> <3>
        <salesforce:records>#[payload]</salesforce:records> <4>
    </salesforce:update>
</flow>
----

> <1> The flow starts with an HTTP Listener that listens for an API call on the `/updatePhoneNumber` path.

Reviewer: `[calloutlist]`?

Are these ones where it doesn't work?

> <2> In the Transform component, the payload is set using the `<ee:transform>` operation to specify the account ID and phone number to update.

Reviewer: I think operations are bold.

> <3> The Salesforce Connector Update operation is used to update the Account object in Salesforce. 

Reviewer: ```suggestion
<3> The Salesforce Connector *Update* operation is used to update the Account object in Salesforce. 
```

+

> The `type` attribute is set to `"Account"` to specify the object type to update.

Reviewer: Do you need both the font change and the ""?

    </salesforce:sfdc-config>
    <flow name="updateAccountPhone">
        <http:listener config-ref="HTTP_Listener_config" doc:name="Listener" path="/updateAccountPhone" allowedMethods="POST" doc:id="listener" />
        <set-variable variableName="accountIdVar" value="#[payload.accountId]" mimeType="application/json" doc:name="Set Account ID var" doc:id="set-variable-accountId" />
        <set-variable variableName="phoneVar" value="#[payload.phoneNumber]" mimeType="application/json" doc:name="Set Phone var" doc:id="set-variable-phone" />
        <salesforce:update type="Account" config-ref="Salesforce_Config" doc:name="Update Salesforce Account" doc:id="update-account">
            <salesforce:records>
                <![CDATA[#[output application/json ---
                [{
                    "Id": vars.accountIdVar,
                    "Phone": vars.phoneVar
                }]]]]>
            </salesforce:records>
        </salesforce:update>
        <logger doc:name="Logger" message="Phone number updated for Account with Id: #[vars.accountIdVar]" doc:id="logger" />
    </flow>
</mule>
----

== Synchronize Account Objects

> This flow enables real-time updates between Salesforce and your database to ensure the data is always current and accurate across both Salesforce and your database. As soon as a change is made in the phone field in Salesforce, it is immediately updated in the database and vice versa.

Reviewer: ```suggestion
This flow enables real-time updates between Salesforce and your database to ensure the data is always current and accurate across both Salesforce and your database. As soon the phone field changes in Salesforce, it updates immediately in the database and vice versa.
```

Reviewer: (not sure if that's correct)

To generate a flow for this task, enter a prompt like: `Create a flow that syncs the phone field of account objects between a Salesforce org and a database`.

> This generates XML code for the new flow, similar to this: 

Reviewer: ```suggestion
This prompt generates XML code for the new flow, similar to this: 
```

Reviewer: Avoid "this" without a noun:
https://confluence.internal.salesforce.com/display/MTDT/MuleSoft+CX+Writing+Style+Reference#MuleSoftCXWritingStyleReference-%22it,%22%22this,%22%22these%22

        <db:sql><![CDATA[SELECT * FROM Account WHERE Id = :accountId]]></db:sql> <4>
        <db:input-parameters><![CDATA[#[{
            "accountId": vars.accountId
        }]]]></db:input-parameters>
    </db:select>
    <ee:transform doc:name="Create Update Data" doc:id="6"> <5>
        <ee:message>
            <ee:set-payload><![CDATA[%dw 2.0
output application/java
---
{
    "Id": vars.accountId,
    "Phone": vars.phone
}]]></ee:set-payload>
        </ee:message>
    </ee:transform>
    <salesforce:update config-ref="Salesforce_Config" doc:name="Update Salesforce Account" doc:id="7" type="Account"> 
        <salesforce:records><![CDATA[#[payload]]]></salesforce:records>
    </salesforce:update> <6>
    <db:update doc:name="Update Account in Database" doc:id="8" config-ref="Database_Config"> 
        <db:sql><![CDATA[UPDATE Account SET Phone = :phone WHERE Id = :accountId]]></db:sql> <7>
        <db:input-parameters><![CDATA[#[{
            "accountId": vars.accountId,
            "phone": vars.phone
        }]]]></db:input-parameters>
    </db:update>
    <set-payload value="#['Phone field synced successfully']" doc:name="Set Payload" doc:id="9" /> <8>
</flow>
----

> <1> The HTTP Listener operation is configured to listen for a POST request on the `/syncAccountPhone` path.

Reviewer: `[calloutlist]` if possible

Reviewer: Any way to make these active voice? Like start all with the verb?

Reviewer: Configure the HTTP Listener...
Extract the account ID...
Extract the ...

Should Listener be bold, since it's an operation?

Global: check that operation names are bold.

> <2> The account ID is extracted from the payload and stored in a variable named `"accountId"`.

Reviewer: Either mono or "" but not both, I think.

ETA: global

<3> The phone number is extracted from the payload and stored in a variable named `"phone"`.
<4> The Database Connector Select operation retrieves the account record from the database based on the account ID.
<5> A DataWeave transformation is used to create the updated data for the Salesforce Account object.

> <6> The Salesforce Connector Update operation udpates the Salesforce account.

Reviewer: ```suggestion
<6> The Salesforce Connector *Update* operation updates the Salesforce account.
```

<7> The account information is also updated in the database using a Database Connector Update operation.
<8> The payload is set to `"Phone field synced successfully"` to indicate a successful sync.

=== Complete the Flow

After reviewing and inserting the generated flow into your project:

> . Update the name of the database table and fields to match what is configured in your MySQL database. 

Reviewer: ```suggestion
. Update the name of the database table and fields to match the configuration in your MySQL database. 
```

Reviewer: (might be incorrect)

        <salesforce:modified-object-listener objectType="Account" config-ref="Salesforce_Config" doc:name="On Modified Object" doc:id="listener">
            <scheduling-strategy>
                <fixed-frequency />
            </scheduling-strategy>
        </salesforce:modified-object-listener>

        <ee:transform doc:name="Transform Message" doc:id="transform">
            <ee:message>
                <ee:set-payload>
                    <![CDATA[%dw 2.0
                    output application/json
                    ---
                    {
                        Id: payload.Id,
                        Phone: payload.Phone
                    }]]>
                </ee:set-payload>
            </ee:message>
        </ee:transform>

        <db:update doc:name="Update Phone in Database" doc:id="update">
            <db:sql>
                <![CDATA[UPDATE Account SET Phone = :Phone WHERE Id = :Id]]>
            </db:sql>
            <db:input-parameters>
                <![CDATA[#[{
                    Phone: payload.Phone,
                    Id: payload.Id
                }]]]>
            </db:input-parameters>
        </db:update>
    </flow>
</mule>
----

> == New Record Notification

Reviewer: Should this be imperative? Automate New Record Notification?

To generate a flow for this task, enter a prompt like: `Generate a flow to send an email when a record is created in the database`.

This generates XML code for the new flow, similar to this: 

[source, xml]
----
<flow name="sendEmailFlow">
    <db:listener table="record" doc:name="On Table Row" doc:id="dbListener" watermarkColumn="CREATION_DATE">
        <scheduling-strategy> <2>
            <fixed-frequency frequency="1" timeUnit="SECONDS"/>
        </scheduling-strategy>
    </db:listener> <1>
    <set-variable variableName="record" value="#[payload]" doc:name="Set Variable" doc:id="setVariable"/> <3>
    <email:send config-ref="Email_SMTP" fromAddress="sender@example.com" subject="New Record Created"> <4>
        <email:to-addresses>
            <email:to-address value="recipient@example.com"/> <5>
        </email:to-addresses>
        <email:body contentType="text/plain">
            <email:content><![CDATA[New record created with ID: #[vars.record.id]]]></email:content>
        </email:body>
    </email:send>
</flow>
----

> <1> The flow starts with a Database Connector Listener component that listens for new records in the database `"record"` table.

Reviewer: `[calloutlist]`

> <2> The scheduling strategy is set to trigger the flow at a fixed frequency of every 1 second.

Reviewer: ```suggestion
<2> The scheduling strategy is set to trigger the flow at a fixed frequency of every one second.
```

Reviewer: (style for 1-9)

> <3> The payload of the listener is stored in a variable named `"record"` using the Set Variable operation.

Reviewer: Listener?

<4> An email is sent from the address `"\sender@example.com"` using the Email Send component.
<5> The email is sent to the address `"\recipient@example.com"`.

=== Complete the Flow

After reviewing and inserting the generated flow into your project:

. Update the sender email address, recipient email addresses, and email body for your specific use case.
. Update the database table name so that it corresponds to the database table on which the flow listens for new records.

> . Update the `watermarkColumn` in the database listener to whichever database column is used to indicate new records. 

Reviewer: Listener?
